From b4cea01d74c6816f6192dc5b6b6c5be77c67f7cd Mon Sep 17 00:00:00 2001
From: Andreas Schmidt <andreas.schmidt@wago.com>
Date: Tue, 6 Nov 2018 08:59:03 +0100
Subject: [PATCH] net: dsa: mv88e6xxx: add wago class device

For abstraction of tcam rules access this patch add a
class device in wago sysfs class.

Signed-off-by: Andreas Schmidt <andreas.schmidt@wago.com>
---
 drivers/net/dsa/mv88e6xxx/chip.c      | 24 +++++++++++++++++++++++-
 drivers/net/dsa/mv88e6xxx/mv88e6xxx.h |  5 +++++
 2 files changed, 28 insertions(+), 1 deletion(-)

diff --git a/drivers/net/dsa/mv88e6xxx/chip.c b/drivers/net/dsa/mv88e6xxx/chip.c
index 42b1c82..1650097 100644
--- a/drivers/net/dsa/mv88e6xxx/chip.c
+++ b/drivers/net/dsa/mv88e6xxx/chip.c
@@ -4383,11 +4383,27 @@ static int mv88e6xxx_probe(struct mdio_device *mdiodev)
 	if (err)
 		goto out_g2_irq;
 
+	chip->class_dev = NULL;
+
+	if (wsysinit_sysfs_class) {
+		struct device *class_dev;
+
+		class_dev = device_create(wsysinit_sysfs_class, chip->dev,
+					  MKDEV(0, 0), chip, "mv88e6xxx_%d",
+					  MINOR(chip->dev->devt));
+		if (IS_ERR(dev)) {
+			err = PTR_ERR(dev);
+			goto out_remove_atu_flush;
+		}
+
+		chip->class_dev = class_dev;
+	}
+
 	INIT_LIST_HEAD(&chip->tcam.info_head.list);
 
 	err = mv88e6xxx_mdio_register(chip, np, false);
 	if (err)
-		goto out_remove_atu_flush;
+		goto out_remove_class_device;
 
 	err = mv88e6xxx_register_switch(chip, np);
 	if (err)
@@ -4405,6 +4421,9 @@ static int mv88e6xxx_probe(struct mdio_device *mdiodev)
 	mv88e6xxx_unregister_switch(chip);
 out_unregister_mdio:
 	mv88e6xxx_mdio_unregister(chip);
+out_remove_class_device:
+	if (chip->class_dev)
+		device_destroy(wsysinit_sysfs_class, chip->class_dev->devt);
 out_remove_atu_flush:
 	device_remove_file(chip->dev, &dev_attr_atu_flush);
 out_g2_irq:
@@ -4425,6 +4444,9 @@ static void mv88e6xxx_remove(struct mdio_device *mdiodev)
 	struct dsa_switch *ds = dev_get_drvdata(&mdiodev->dev);
 	struct mv88e6xxx_chip *chip = ds->priv;
 
+	if (chip->class_dev)
+		device_destroy(wsysinit_sysfs_class, chip->class_dev->devt);
+
 	device_remove_file(chip->dev, &dev_attr_atu_flush);
 	mv88e6xxx_tcam_cleanup(chip);
 	mv88e6xxx_debugfs_destroy(chip);
diff --git a/drivers/net/dsa/mv88e6xxx/mv88e6xxx.h b/drivers/net/dsa/mv88e6xxx/mv88e6xxx.h
index abdce58..7de6918 100644
--- a/drivers/net/dsa/mv88e6xxx/mv88e6xxx.h
+++ b/drivers/net/dsa/mv88e6xxx/mv88e6xxx.h
@@ -641,6 +641,8 @@ enum mv88e6xxx_cap {
 	 MV88E6XXX_FLAGS_PVT |		\
 	 MV88E6XXX_FLAGS_SERDES)
 
+extern struct class *wsysinit_sysfs_class;
+
 struct mv88e6xxx_ops;
 
 struct mv88e6xxx_info {
@@ -709,6 +711,9 @@ struct mv88e6xxx_chip {
 	/* The device this structure is associated to */
 	struct device *dev;
 
+	/* Wago sysfs class device */
+	struct device *class_dev;
+
 	/* This mutex protects the access to the switch registers */
 	struct mutex reg_lock;
 
-- 
2.7.4

